### PR TITLE
Make clusterctl provider agnostic / remove deployer

### LINKED ISSUE #1040: Deprecate provider.Deployer
/kind cleanup
/area clusterctl
/milestone v1alpha2

**Describe the solution you'd like**
This issue is to track the development towards v1alpha2, including [Machine States & Preboot Bootstrap Proposal](https://github.com/kubernetes-sigs/cluster-api/pull/997).

The [provider.Deployer interface](https://github.com/kubernetes-sigs/cluster-api/blob/release-0.1/cmd/clusterctl/clusterdeployer/provider/provider.go) has had the deprecation notice for quite some time. 

As part of v1alpha2 changes, the immediate deprecation of providerSpec, actuators, and the introduction of secret-based KubeConfig to gain access to remote clusters, we can now aim to deprecate this interface in this iteration.

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1040

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

/assign @timothysc

### CODE REVIEW

**Comment by pablochacin** [comment]
> This looks fine to me, but with the discussion moving towards an operator and separate tooling, should we instead just remove `clusterctl` all together?

Isn't there still value in using `clusterctl` for orchestrating the process, even if it could be done "manually" by creating all the required CRDs?

**Comment by pablochacin** [comment]
`cmd/clusterclt/README.md` should be updated:

>Due to the limitations described below, you must currently compile and run a clusterctl binary from your chosen provider implementation rather than using the binary from this repository.

Also
>Limitations
>clusterctl can only use a provider that is compiled in. As provider specific code has been moved out of this repository, running the clusterctl binary compiled from this repository isn't particularly useful.
>
>There is current work ongoing to rectify this issue, which centers around removing the ProviderDeployer interface from the clusterdeployer package. The two tracking issues for removing the two functions in the interface are https://github.com/kubernetes-sigs/cluster-api/issues/158 and https://github.com/kubernetes-sigs/cluster-api/issues/160.~~

**Comment by detiber** [comment]
> > This looks fine to me, but with the discussion moving towards an operator and separate tooling, should we instead just remove clusterctl all together?

> Isn't there still value in using clusterctl for orchestrating the process, even if it could be done "manually" by creating all the required CRDs?

Yes, there is still value in having a tool to orchestrate the process, but @timothysc was suggesting that the tool live alongside the `cluster-api-operator` in a separate repo that could be versioned and documented specifically towards end users.

File: cmd/clusterctl/phases/getkubeconfig.go [hunk 1]
```diff
@@ -62,18 +61,14 @@ func waitForKubeconfigReady(bootstrapClient clusterclient.Client, provider provi
...
-		klog.V(2).Infof("Waiting for kubeconfig on %v to become ready...", controlPlane.Name)
-		k, err := provider.GetKubeConfig(cluster, controlPlane)
+		klog.V(2).Infof("Waiting for kubeconfig from Secret %q-kubeconfig in namespace %q to become available...", clusterName, namespace)
+		k, err := bootstrapClient.GetKubeconfigFromSecret(namespace, clusterName)
 		if err != nil {
 			klog.V(4).Infof("error getting kubeconfig: %v", err)
 			return false, nil
 		}
 		if k == "" {
+			klog.V(4).Info("error getting kubeconfig: secret value is empty")
```
Comment (timothysc) [inline cmd/clusterctl/phases/getkubeconfig.go]: Shouldn't an empty secret be an error from the main function?  If so why not?

File: cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go [hunk 2]
```diff
@@ -142,6 +140,20 @@ func (c *client) EnsureNamespace(namespaceName string) error {
...
 }
 
+func (c *client) GetKubeconfigFromSecret(namespace, clusterName string) (string, error) {
+	clientset, err := clientcmd.NewCoreClientSetForDefaultSearchPath(c.kubeconfigFile, clientcmd.NewConfigOverrides())
+	if err != nil {
+		return "", errors.Wrap(err, "error creating core clientset")
+	}
+
+	secretName := fmt.Sprintf("%s-kubeconfig", clusterName)
+	secret, err := clientset.CoreV1().Secrets(namespace).Get(secretName, metav1.GetOptions{})
```
Comment (timothysc) [inline cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go]: Why is there no error checking on the return data.  e.g. empty string, etc?

**Comment by CamelCaseNotation** [comment]
Does https://cluster-api.sigs.k8s.io/provider_implementations/register_schemes.html need to be updated due to this commit? Specifically the comment:
`// Register our cluster deployer (the interface is in clusterctl and we define the Deployer interface on the actuator)`

**Comment by CamelCaseNotation** [comment]
https://cluster-api.sigs.k8s.io/provider_implementations/create_actuators.html
`GetIP` and `GetKubeConfig` no longer need to be implemented, correct? Where can I update this documentation?

**Comment by vincepri** [comment]
Yes the documentation needs to be updated for v1alpha2

**Pull request merged**